File: masker/masking.py
"""Masking & tokenization. CODEX OWNS THIS FILE.

Replaces sensitive entity spans with either a placeholder (`[MASKED:type]`)
or a stable token (`<TOKEN:abcd1234>`). Returns a MaskedText that includes
a token map so output filtering can re-mask leaked values on the way back.

Contract:
    mask(text: str, detection: DetectionResult, *, mode: str) -> MaskedText
    unmask(text: str, masked: MaskedText) -> str
"""

from __future__ import annotations

import hashlib
from typing import Literal

from .contracts import DetectionResult, EntityType, MaskedText

MaskMode = Literal["placeholder", "token"]


def _token_for(value: str) -> str:
    digest = hashlib.sha256(value.encode("utf-8")).hexdigest()[:8]
    return f"<TOKEN:{digest}>"


def mask(
    text: str,
    detection: DetectionResult,
    *,
    mode: MaskMode = "placeholder",
) -> MaskedText:
    """Replace each detected entity span in `text`.

    `placeholder` is more readable for the LLM (`[MASKED:ssn]`).
    `token` is reversible and useful when the LLM needs to refer back
    to the same redacted entity multiple times.
    """
    spans = sorted(
        [
            e
            for e in detection.entities
            if e.start >= 0 and e.end > e.start and e.type != EntityType.HEALTH_CONTEXT
        ],
        key=lambda e: e.start,
        reverse=True,
    )

    out = text
    token_map: dict[str, str] = {}
    replacements: list[dict[str, str | int]] = []
    occupied: list[tuple[int, int]] = []
    for e in spans:
        if any(not (e.end <= start or e.start >= end) for start, end in occupied):
            continue
        if mode == "token":
            replacement = _token_for(e.value)
            token_map[replacement] = e.value
        else:
            replacement = "[MASKED]"
        out = out[: e.start] + replacement + out[e.end :]
        replacements.append(
            {
                "type": e.type.value,
                "original": e.value,
                "replacement": replacement,
                "start": e.start,
                "end": e.end,
            }
        )
        occupied.append((e.start, e.end))

    replacements.reverse()
    return MaskedText(text=out, token_map=token_map, replacements=replacements)


def unmask(text: str, masked: MaskedText) -> str:
    """Inverse of `mask` for `token` mode. For `placeholder` mode the original
    values are restored as well, but readers should be aware that placeholders
    in LLM output may not exactly match the originals.
    """
    out = text
    for placeholder, value in masked.token_map.items():
        out = out.replace(placeholder, value)
    return out


def scrub_output(text: str, detection: DetectionResult) -> str:
    """Re-run masking on a model output to catch any leaked sensitive
    values the LLM might have echoed back in unredacted form.
    """
    out = text
    for e in detection.entities:
        if not e.value:
            continue
        out = out.replace(e.value, "[MASKED]")
    return out

File: masker-rs/crates/masker/src/masking.rs
//! Masking & tokenization. Mirrors `masker/masking.py`.

use std::collections::BTreeMap;

use sha2::{Digest, Sha256};

use crate::contracts::{DetectionResult, MaskedText};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MaskMode {
    /// `[MASKED:type]` — readable for the LLM.
    Placeholder,
    /// `<TOKEN:abcd1234>` — stable handle so the LLM can refer back.
    Token,
}

fn token_for(value: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(value.as_bytes());
    let digest = hasher.finalize();
    let hex = hex::encode(digest);
    format!("<TOKEN:{}>", &hex[..8])
}

/// Replace each detected entity span in `text`.
pub fn mask(text: &str, detection: &DetectionResult, mode: MaskMode) -> MaskedText {
    let mut spans: Vec<(usize, usize, String, &'static str)> = detection
        .entities
        .iter()
        .filter(|e| e.end > e.start && e.start < text.len() && e.end <= text.len())
        .map(|e| (e.start, e.end, e.value.clone(), e.kind.as_str()))
        .collect();
    spans.sort_by_key(|(s, _, _, _)| std::cmp::Reverse(*s));

    let mut out = text.to_string();
    let mut token_map: BTreeMap<String, String> = BTreeMap::new();

    for (start, end, value, kind) in spans {
        let replacement = match mode {
            MaskMode::Token => token_for(&value),
            MaskMode::Placeholder => format!("[MASKED:{}]", kind),
        };
        token_map.insert(replacement.clone(), value);

        if !out.is_char_boundary(start) || !out.is_char_boundary(end) {
            continue;
        }
        out.replace_range(start..end, &replacement);
    }

    MaskedText {
        text: out,
        token_map,
    }
}

/// Inverse of [`mask`] — substitute placeholders/tokens back to originals.
pub fn unmask(text: &str, masked: &MaskedText) -> String {
    let mut out = text.to_string();
    for (placeholder, value) in &masked.token_map {
        out = out.replace(placeholder.as_str(), value);
    }
    out
}

/// Re-mask any detected entity values that leaked verbatim into model output.
pub fn scrub_output(text: &str, detection: &DetectionResult) -> String {
    let mut out = text.to_string();
    for e in &detection.entities {
        if e.value.is_empty() {
            continue;
        }
        let replacement = format!("[MASKED:{}]", e.kind.as_str());
        out = out.replace(e.value.as_str(), &replacement);
    }
    out
}
